### PR TITLE
Pin to numpy<1.25 and numba<0.58 to avoid errors and deprecation warnings-as-errors.

### LINKED ISSUE #14155: [BUG] Pin wheel builds to `numpy <1.25`
**Describe the bug**

Numpy 1.26 is out, and produces deprecation warnings in pandas (1.5.x). We notice this in wheel builds which don't have an upper bound pin (e.g. https://github.com/rapidsai/cudf/actions/runs/6262541728/job/17005833818?pr=14142)

**Expected behavior**

Pin to appropriate version

### PR DESCRIPTION
**Pull request by bdice**
## Description
Closes #14155.

Related: #14160.

(Will newer numpy support be backported to pandas 1.x? edit: no, see below)

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by mroeschke** [comment]
Sorry, `np.find_common_type` usage was only addressed in 2.0.x and won't be backported to 1.x: https://github.com/pandas-dev/pandas/pull/53343

File: conda/recipes/cudf/meta.yaml [hunk 1]
```diff
@@ -79,7 +79,8 @@ requirements:
     - pandas >=1.3,<1.6.0dev0
     - cupy >=12.0.0
     - numba >=0.57
-    - numpy >=1.21
+      # TODO: Pin to numpy <1.25 until upgrading to pandas 2
```
Comment (bdice) [inline conda/recipes/cudf/meta.yaml:82]:
```suggestion
    # TODO: Pin to numpy <1.25 until upgrading to pandas 2
```

**Review by vyasr — APPROVED** [review]

**Comment by vyasr** [comment]
@bdice do you want to create another issue like #14160 for the numpy warnings (which we will also need to fix)? Or repurpose #14155 and leave it open after this PR merges, either is fine with me as long as we're tracking the work.

**Review by quasiben — APPROVED** [review]

**Review by raydouglass — APPROVED** [review]

**Review by galipremsagar — APPROVED** [review]

**Comment by bdice** [comment]
> @bdice do you want to create another issue like #14160 for the numpy warnings (which we will also need to fix)? Or repurpose #14155 and leave it open after this PR merges, either is fine with me as long as we're tracking the work.

I don't have a strong opinion about how we resolve this -- maybe a new issue is best. But for now, I don't see a lot of benefit to fixing these warnings until we are ready to migrate cudf to pandas 2. Then it'll be easier to tell what's broken in cudf and not just what's broken in pandas internals. If you want to figure that out, I'd appreciate help with filing a second issue (my plate is very full for today already).

**Comment by vyasr** [comment]
> But for now, I don't see a lot of benefit to fixing these warnings until we are ready to migrate cudf to pandas 2. Then it'll be easier to tell what usage is deprecated in cudf and not just what's deprecated in the pandas internals. If you want to figure that out, I'd appreciate help with filing a second issue (my plate is very full for today already).

I agree, I don't think we should work on this immediately. I just want an issue tracking this so that we don't forget to move the pin when we bump our pandas version. I'll open a stub issue for tracking that we can fill out with more info later.

**Comment by vyasr** [comment]
Opened https://github.com/rapidsai/cudf/issues/14166

**Pull request merged**